modules/adagioBidAdapter.js

```diff
@@ -1,73 +1,355 @@
- * - replaced `mobile` by `mobi` in the table regexp, so Opera Mobile on phones is not detected as a tablet.
- */
-function _getDeviceType() {
-  let ua = navigator.userAgent;
+function canAccessTopWindow() {
+  try {
+    if (utils.getWindowTop().location.href) {
+      return true;
+    }
+  } catch (error) {
+    return false;
+  }
+}
+
+function initAdagio() {
+  const script = document.createElement('script');
+  const w = utils.getWindowTop();
+
+  w.ADAGIO = w.ADAGIO || {};
+  w.ADAGIO.queue = w.ADAGIO.queue || [];
+  w.ADAGIO.versions = w.ADAGIO.versions || {};
+  w.ADAGIO.versions.adagioBidderAdapter = VERSION;
+
+  const getAdagioTag = function getAdagioTag() {
+    const ls = w.localStorage.getItem('adagioScript');
+    if (ls !== null) {
+      Function(ls)(); // eslint-disable-line no-new-func
+    } else {
+      utils.logWarn('Adagio Script not found');
+    }
+  }
+
+  const adagioEnqueue = function adagioEnqueue(action, data) {
+    w.ADAGIO.queue.push({ action, data, ts: Date.now() });
+  }
+
+  // Listen to ad-server events in current window as we can
+  // in a Post-Bid scenario.
+  window.googletag = window.googletag || {};
+  window.googletag.cmd = window.googletag.cmd || [];
+  window.googletag.cmd.push(function() {
+    const gptEvents = Object.keys(ADSRV_EVENTS.GPT).map(key => ADSRV_EVENTS.GPT[key]);
+    gptEvents.forEach(eventName => {
+      window.googletag.pubads().addEventListener(eventName, args => {
+        adagioEnqueue('gpt-event', { eventName, args });
+      });
+    });
+  });
+
+  window.sas = window.sas || {};
+  window.sas.cmd = window.sas.cmd || [];
+  window.sas.cmd.push(function() {
+    const sasEvents = Object.keys(ADSRV_EVENTS.SAS).map(key => ADSRV_EVENTS.SAS[key]);
+    sasEvents.forEach(eventName => {
+      window.sas.events.on(eventName, args => {
+        adagioEnqueue('sas-event', { eventName, args });
+      });
+    });
+  });
 
-  // Tablets must be checked before phones.
-  if ((/(tablet|ipad|playbook|silk)|(android(?!.*mobi))/i).test(ua)) {
-    return 5; // "tablet"
+  if (window.advstLib && typeof window.advstLib === 'function') {
+    advstLib().eventHandler.addListener('renderEndedEvent', function(e) {
+      adagioEnqueue('adthk-event', { eventName: 'renderEndedEvent', args: e.detail });
+    });
   }
-  if ((/Mobile|iP(hone|od|ad)|Android|BlackBerry|IEMobile|Kindle|NetFront|Silk-Accelerated|(hpw|web)OS|Fennec|Minimo|Opera M(obi|ini)|Blazer|Dolfin|Dolphin|Skyfire|Zune/).test(ua)) {
-    return 4; // "phone"
+
+  // First, try to load adagio-js from localStorage.
+  getAdagioTag();
+
+  // Then prepare localstore.js to update localStorage adagio-sj script with
+  // the very last version.
+  script.type = 'text/javascript';
+  script.async = true;
+  script.src = ADAGIO_TAG_URL;
+  script.setAttribute('data-key', ADAGIO_LOCALSTORE_KEY);
+  script.setAttribute('data-src', ADAGIO_TAG_TO_LOCALSTORE);
+  setTimeout(function() {
+    utils.insertElement(script);
```

Comment:
Please use the `loadExternalScript` function instead of directly inserting the javascript. 
See https://github.com/prebid/Prebid.js/blob/master/src/adloader.js#L15 (You'll need to add your bidder to the whitelist).

Reply:
Done, see d9c2720 and c0f2520